lib/Doctrine/DBAL/Driver/PDOConnection.php

```diff
@@ -104,11 +104,13 @@ public function quote($value, $type = ParameterType::STRING)
     public function lastInsertId($name = null)
     {
         try {
-            if ($name === null) {
-                return parent::lastInsertId();
-            }
+            $lastInsertId = $name === null
+                ? parent::lastInsertId()
+                : parent::lastInsertId($name);
 
-            return parent::lastInsertId($name);
+            assert($lastInsertId !== false);
```

Comment:
I was unsure how to fix this one. According to the documentation, `PDO::lastInsertId()` might return `false`. My assumption is that this can only happen if the PDO instance is configured to not raise exceptions (which is not the case in our context).

However, on the 3.1.x branch, Psalm will complain about this assertion as being redundant, see #4944.

Reply:
> My assumption is that this can only happen if the PDO instance is configured to not raise exceptions

I'd expect it to return false if there's no last insert available? E.g. on a brand-new connection. See https://github.com/doctrine/dbal/blob/4b1cce0abc559d6b93c49f33f7c6a3b4fe9103f7/tests/Functional/WriteTest.php#L203

Reply:
> > My assumption is that this can only happen if the PDO instance is configured to not raise exceptions
> 
> I'd expect it to return false if there's no last insert available? E.g. on a brand-new connection.

Hmm, no. With `pdo_mysql` and `pdo_sqlite` I get the string `0` in that case. With `pdo_pgsql`, I get an exception:

```
Uncaught PDOException: SQLSTATE[55000]: Object not in prerequisite state: 7 ERROR:  lastval is not yet defined in this session
```

However, if I set the PDO error mode to silent, I actually get `false` as return value.

> See
> 
> https://github.com/doctrine/dbal/blob/4b1cce0abc559d6b93c49f33f7c6a3b4fe9103f7/tests/Functional/WriteTest.php#L203

This test is skipped for all the drivers I can test locally. Is this test executed anywhere? Oracle maybe?

Reply:
Yes, oci8. But actually, it's not the test I wanted to point out. This: https://github.com/doctrine/dbal/blob/824eb7ddda50d204a49a4220087056a0a68cbfdc/tests/Functional/WriteTest.php#L169-L180

Also see https://github.com/doctrine/dbal/pull/4885#issuecomment-945019095.

Reply:
Okay, so how should I proceed? Should our version of `lastInsertId()` return `false` if the PDO drivers does so? In that case, our doc blocks are wrong and my newly added `assert()` would do more harm than good.

Shall I revert my changes to this method and have PHPStan ignore the error, so we can deal with it properly on a higher branch?

Reply:
I believe until `4.0.x` where this method starts throwing an exception, we should return the value from PDO as is. Therefore, document the method as returning `false` and revert the assertion. By being just a PDO wrapper in `2.x` and partially in `3.x`, the DBAL cannot guard its API consumers against the underlying API changes.